src/Microsoft.AspNet.OData.Shared/Query/Expressions/DynamicTypeWrapper.cs

```diff
@@ -123,5 +128,65 @@ internal class NoGroupByAggregationWrapper : GroupByWrapper
 
     internal class EntitySetAggregationWrapper : GroupByWrapper
     {
-    }    
+    }
+
+    internal class ComputeWrapper<T> : GroupByWrapper, IEdmEntityObject
+    {
+        public T Instance { get; set; }
+
+        /// <summary>
+        /// An ID to uniquely identify the model in the <see cref="ModelContainer"/>.
+        /// </summary>
+        public string ModelID { get; set; }
+
+        private bool _merged;
+        protected override void EnsureValues()
+        {
+            base.EnsureValues();
+            if (!this._merged)
+            {
+                // Base properties availbale via Instance can be real OData properties or generated in previous transformations
+
+                var instanceContainer = this.Instance as DynamicTypeWrapper;
+                if (instanceContainer != null)
+                {
+                    // Add proeprties generated in previous transformations to the collection
+                    _values.MergeWithReplace(instanceContainer.Values);
+                }
+                else
+                {
+                    // Add real OData properties to the collection
+                    // We need to use injected Model to real proeprty names
+                    var edmType = GetEdmType() as IEdmEntityTypeReference;
```

Comment:
>IEdmEntityTypeReference [](start = 50, length = 23)

Will this cast always succeed, or might GetEdmType return something besides an IEdmEntityTypeReference?
What about complex types?

Reply:
For this PR (adding compute() to $apply), it should be true all the time, because we are using only ```EntitySet?$apply=....```

When we will work on $apply inside $expand (I have that feature in my fork). For queries like:
```EntitySet?$expand=ComplexTypeCollection($apply=...)``` it will change.  I think it will be really difficult to validate that ComputeWrapper works for complex types new before implementing $expand($apply) and/or $expand($compute)